Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -88,14 +88,57 @@ extension HTTPClient {
 
     /// Represent HTTP request.
     public struct Request {
+
+        /// Represent kind of Request
+        enum Kind {
+            /// Remote host request.
+            case host
+            /// UNIX Domain Socket HTTP request.
+            case unixSocket
+
+            private static var hostSchemes = ["http", "https"]
+            private static var unixSchemes = ["unix"]
+
+            init(forScheme scheme: String) throws {
+                if Self.host.supports(scheme: scheme) {
+                    self = .host
+                } else if Self.unixSocket.supports(scheme: scheme) {
+                    self = .unixSocket
+                } else {
+                    throw HTTPClientError.unsupportedScheme(scheme)
+                }
+            }
+
+            func hostFromURL(_ url: URL) throws -> String {
+                switch self {
+                    case .host:
+                        guard let host = url.host else {
+                            throw HTTPClientError.emptyHost
+                        }
+                        return host
+                    case .unixSocket:
+                        return ""
+                }
+            }
+
+            func supports(scheme: String) -> Bool {
+                switch self {
+                case .host:
+                    return Self.hostSchemes.contains(scheme)
+                case .unixSocket:
+                    return Self.unixSchemes.contains(scheme)
+                }
+            }
+        }
+
         /// Request HTTP method, defaults to `GET`.
-        public let method: HTTPMethod
+        public var method: HTTPMethod
         /// Remote URL.
-        public let url: URL
+        public var url: URL
         /// Remote HTTP scheme, resolved from `URL`.
-        public let scheme: String
+        public var scheme: String
         /// Remote host, resolved from `URL`.
-        public let host: String
+        public var host: String
```

Comment:
why do we need those properties to be mutable? If someone changes their value we have to re-check all our invariants (scheme, host, kind, etc)

Reply:
we don't. I changed it when it was computed value, and didn't change it back when reverted it to the property

Reply:
reverted 535303b